gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java

```diff
@@ -717,24 +717,26 @@ public void run() {
 
       try {
         if (!oldEntryMap.isEmpty()) {
-          for (Map<Object, BlockingQueue<RegionEntry>> regionEntryMap : oldEntryMap.values()) {
+          for (Entry<String,Map<Object, BlockingQueue<RegionEntry>>> entry : oldEntryMap.entrySet()) {
+            Map<Object, BlockingQueue<RegionEntry>> regionEntryMap = entry.getValue();
+            Region region = getRegion(entry.getKey());
             for (BlockingQueue<RegionEntry> oldEntriesQueue : regionEntryMap.values()) {
               for (RegionEntry re : oldEntriesQueue) {
                 boolean entryFoundInTxState = false;
                 for (TXStateProxy txProxy : getTxManager().getHostedTransactionsInProgress()) {
                   TXState txState = txProxy.getLocalTXState();
-                  if (txState != null && txState.containsRegionEntryReference(re) && !txState
-                      .isCommitted()) {
+                  if (txState != null && txState.checkEntryVersion(region, re) && !txState
+                      .isCommitted() && re.isUpdateInProgress()) {
```

Comment:
The condition should be
if(re.isUpdateInProgress() || txState != null && !txState
                      .isCommitted() && txState.checkEntryVersion(region, re))